OSDOCS#4348: Installing an IBM Cloud VPC cluster in restricted env

mjpytlak opened the pull request:
Version(s):
4.15+

Issue:
This issue addresses [osdocs-4348](https://issues.redhat.com/browse/OSDOCS-4348).

Link to docs preview:
[Installing a cluster on IBM Cloud VPC in a restricted network](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html)

A significant amount of the content in this topic is reused from existing approved IBM Cloud VPC doc. The following preview links are for areas where notable new content was added in support of a restricted installation:

- [Supported installation methods for different platforms](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing-preparing#supported-installation-methods-for-different-platforms) (Updated support matrix to include a restricted installation)
- [Prerequisites](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#prerequisites_installing-ibm-cloud-restricted) (updated to list using oc-mirror to mirror install media as a prereq)
- [Required internet access and an installation host](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#required-internet-access-and-an-installation-host_installing-ibm-cloud-restricted)
- [Access to IBM service endpoints](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#access-to-ibm-service-endpoints_installing-ibm-cloud-restricted)
- [Allowing endpoint gateway traffic](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#installation-ibm-cloud-configure-vpc-for-endpoint-gateways_installing-ibm-cloud-restricted)
- [Downloading the RHCOS cluster image](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#installation-ibm-cloud-download-rhcos_installing-ibm-cloud-restricted)
- [Manually creating the installation configuration file](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#installation-initializing-manual_installing-ibm-cloud-restricted)
- [Sample customized install-config.yaml file for IBM Cloud VPC](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#installation-ibm-cloud-config-yaml_installing-ibm-cloud-restricted) (added private endpoints as 9th annotation)
- [Disabling the default OperatorHub catalog sources](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#olm-restricted-networks-operatorhub_installing-ibm-cloud-restricted)
- [Installing the policy resources into the cluster](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#oc-mirror-updating-cluster-manifests_installing-ibm-cloud-restricted)
- [Additional IBM Cloud configuration parameters](https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installation-config-parameters-ibm-cloud-vpc#installation-configuration-parameters-additional-ibm-cloud_installation-config-parameters-ibm-cloud-vpc) (added serviceEndpoints)


QE review:
- [x] QE has approved this change.

mjpytlak commented:
@MayXuQQ @jeffnowicki Ready for SME and QE review. PTAL. Thanks in advance.

mjpytlak commented on modules/installation-about-restricted-network.adoc:
```diff
@@ -37,6 +38,10 @@ endif::[]
+ifeval::["{context}" == "installing-ibm-cloud-restricted"]
+:ipi:
+endif::[]
+
```
@jeffnowicki Line 56 makes reference to AWS makes reference to services that still require internet access. I would like to add content that is specific/relevant to IBM Cloud VPC. Any input on what would still require internet access. Thanks.

kalexand-rh commented:
/remove-lifecycle stale

mjpytlak commented:
Thanks for removing that label @kalexand-rh I did not receive a notification about the state. This work has pushed to 4.15.

MayXuQQ commented:
/remove-lifecycle stale

MayXuQQ commented:
https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing-preparing#supported-installation-methods-for-different-platforms
Table 2. User-provisioned infrastructure options
IBM Cloud
- Custom 
- Network customization
- Shared VPC hosted outside of cluster project
cc @jeffnowicki

mjpytlak commented:
> Table 2. User-provisioned infrastructure options
> IBM Cloud

We do not support a UPI installation on IBM Cloud @MayXuQQ

MayXuQQ commented:
https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#installation-ibm-cloud-config-yaml_installing-ibm-cloud-restricted
currently, we just support us-east, since the VPE limitation. cc @jianlinliu @jeffnowicki 

> OCP install in a disconnected network need to access https://private.resource-controller.cloud.ibm.com/, while the url is only available in some limited region (e.g: us-east) after VPE created, that would be a big limitation for customers. A common IPI can be created in all regions, while disconnected install is only supported in some specific region, that would be a bit hard to understand for customers.

mjpytlak commented:
VPEs are not regionally bound. In the sample config file you reference we state that specifying private endpoints in the same region, or geographically near to the region the cluster is being deployed to, is preferable to minimize latency.

MayXuQQ commented:
> I originally had the same concern, but confirmed with @jeffnowicki that VPEs are not regionally bound. In the sample config file you reference we state that specifying private endpoints in the same region, or geographically near to the region the cluster is being deployed to, is preferable to minimize latency.

based on the current test, now just can be tested in us-east region. 
cc @jianlinliu

jianlinliu commented:
> VPEs are not regionally bound. In the sample config file you reference we state that specifying private endpoints in the same region, or geographically near to the region the cluster is being deployed to, is preferable to minimize latency.

This was our [discussion](https://redhat-internal.slack.com/archives/C066ML5GY0J/p1701082943409679?thread_ts=1700596002.427439&cid=C066ML5GY0J) about accessing private endpoints.

> for https://private.resource-controller.cloud.ibm.com/, my cluster is being installed in eu-gb, but from the output of ibmcloud is endpoint-gateway-targets seem like there is no such endpoint target is listed, but if I run the command against us-east  , it can be listed, does that mean the disconnected install is only supported in us-east?

> Yes, I think that is likely the case, that only a certain set of IBM Cloud Regions will support Disconnected. But, there is some disagreement on our team still on whether using global endpoints will allow Disconnected support in all Regions. So far, I have not yet attempted different configurations, but I have had to migrate the IBM Cloud Region I started Disconnected testing with a few times until I reached us-east. Unfortunately, the documentation for different IBM Cloud Services makes it difficult to piece together all of the limitations of each IBM Cloud Service to find the proper Region to use.

When creating VPEs, firstly users have to switch to the region he/she are going to install the cluster, in my testing, once I switch to non-us-east region, there would be no any resource-controller endpoint to access. Standing from customers, how to move forwards the installation in such region?

jianlinliu commented on modules/installation-ibm-cloud-config-yaml.adoc:
```diff
@@ -283,6 +287,143 @@ For production {product-title} clusters on which you want to perform installatio
+platform:
+  ibmcloud:
+    region: us-east <1>
+    resourceGroupName: us-east-example-network-rg <8>
```
I do not think this is a MUST, and `resourceGroupName` is used to specify where the cluster is deployed to, not have to share with VPC resources.

jianlinliu replied:
```suggestion
    resourceGroupName: us-east-example-cluster-rg <8>
```

jianlinliu commented on modules/installation-ibm-cloud-config-yaml.adoc:
```diff
@@ -283,6 +287,143 @@ For production {product-title} clusters on which you want to perform installatio
+        url: https://private.us-east.resource-controller.cloud.ibm.com
+      - name: ResourceManager
+        url: https://private.us-east.resource-controller.cloud.ibm.com
+      - name: DNSServices
```
we only support DNSSerivce endpoint override, but not CIS, so the IBMCloud install in a restrict network has to be a private cluster together, right? But the doc, it is saying it is optional.
```
Optional: Set the publishing strategy to `Internal`:
```

jeffnowicki replied:
> we only support DNSSerivce endpoint override, but not CIS, so the IBMCloud install in a restrict network has to be a private cluster together, right? But the doc, it is saying it is optional.
> 
> ```
> Optional: Set the publishing strategy to `Internal`:
> ```

You could use a publishing strategy of `external` for a restricted/disconnected network install. In which case, a private service endpoint override for CIS would be needed. This use case seems odd, but theoretically possible.

jianlinliu replied:
>  I need confirm my statement and verify availability of a private CIS service endpoint.

On https://cloud.ibm.com/vpc-ext/provision/endpointGateway, I only can see a private DNSService endpoint, but no CIS.

jianlinliu commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.adoc:
```diff
@@ -0,0 +1,82 @@
++
+[NOTE]
+====
+Because the installation media is on the mirror host, you can use that computer to complete all installation steps.
```
Once overriding the private endpoints the install process, the installer has to be able to access the private endpoints from the host where installer is running for validation, so that means it is requirements to ensure installer can access these private endpoints.

mjpytlak replied:
If I am following you, I believe my latest updates to  About installations in restricted networks > Required internet access and an installation host [1] should address your concern here.

[1]  https://github.com/openshift/openshift-docs/pull/55526/commits/1db5205bc4277baf39523a3bd249b4d7d2a62ea1#diff-0995ccf569e68d8a5e211b07d6f6bd89d741f4761fb517528c10bf7b445f9593R84

jianlinliu replied:
Can we add one more item under "Required internet access and an installation host" and state:

* Use the installation program to access the overriding private endpoints.

mjpytlak replied:
IMO I believe that we are covered when we state internet access is required to use the installation program to create the installation file. Users are instructed to use private endpoints as part of that procedure [1]. Is there a reason that you believe it should be explicitly called out?

[1] https://github.com/openshift/openshift-docs/commit/1db5205bc4277baf39523a3bd249b4d7d2a62ea1#diff-aa2bf98d200d7bbc6fbc805b89a39df30e54112c533f035e025e1160cfeeecd9R607

jianlinliu replied:
On IBMCloud, once any endpoint is overridden by a private url, the installation program has to be run inside VPC, even it is not a OCP installation in restricted network.

The explicit description may help users understand. If you think the current description, I am okay with it, too,

mjpytlak commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.adoc:
```diff
@@ -0,0 +1,82 @@
+** Has firewall rules or a peering connection to access the mirror registry that is hosted elsewhere.
+
++
+You cannot use the VPC that the installation program provisions by default.
```
As part of our conversation, we agreed to leave the current statement as is. There is no need to cross reference "Installing a cluster on IBM Cloud into an existing VPC". All of the content in that topic is shared in this topic.

mjpytlak commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.adoc:
```diff
@@ -0,0 +1,82 @@
+:_content-type: ASSEMBLY
+[id="installing-ibm-cloud-restricted"]
+= Installing a cluster on IBM Cloud VPC in a restricted network
```
Done. Thank you.

mjpytlak commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.adoc:
```diff
@@ -0,0 +1,82 @@
+
+toc::[]
+
+In {product-title} {product-version}, you can install a cluster in a restricted network by creating an internal mirror of the installation release content that is accessible to an existing Virtual Private Cloud (VPC) on {ibm-cloud-name}.
```
Thanks again for this, Jeff. As discussed, rather than add it to the abstract of the topic, I updated About installations in restricted networks to address each of the points you highlighted.

mjpytlak commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.adoc:
```diff
@@ -0,0 +1,82 @@
+* You reviewed details about the xref:../../architecture/architecture-installation.adoc#architecture-installation[{product-title} installation and update] processes.
+* You read the documentation on xref:../../installing/installing-preparing.adoc#installing-preparing[selecting a cluster installation method and preparing it for users].
+* You xref:../../installing/installing_ibm_cloud_public/installing-ibm-cloud-account.adoc#installing-ibm-cloud-account[configured an IBM Cloud account] to host the cluster.
+* You have a container image registry, such as Red Hat Quay. If you do not already have a registry, you can create a mirror registry using  xref:../../installing/disconnected_install/installing-mirroring-creating-registry.adoc#installing-mirroring-creating-registry[_mirror registry for Red Hat OpenShift_].
```
Agreed. I consolidated the prerequisite statements into one.

mjpytlak commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.adoc:
```diff
@@ -0,0 +1,82 @@
++
+You cannot use the VPC that the installation program provisions by default.
+* If you use a firewall, you xref:../../installing/install_config/configuring-firewall.adoc#configuring-firewall[configured it to allow the sites] that your cluster requires access to.
+* You configured the `ccoctl` utility before you installed the cluster. For more information, see xref:../../installing/installing_ibm_cloud_public/configuring-iam-ibm-cloud.adoc#configuring-iam-ibm-cloud[Configuring IAM for IBM Cloud VPC].
```
Per our conversation, we agreed to leave this content as is for now. It is existing content that is currently managed by another writer.

mjpytlak commented on modules/installation-ibm-cloud-config-yaml.adoc:
```diff
@@ -283,6 +287,143 @@ For production {product-title} clusters on which you want to perform installatio
+<5> The machine CIDR must contain the subnets for the compute machines and control plane machines.
+<6> The CIDR must contain the subnets defined in `platform.ibmcloud.controlPlaneSubnets` and `platform.ibmcloud.computeSubnets`.
+<7> The cluster network plugin to install. The supported values are `OVNKubernetes` and `OpenShiftSDN`. The default value is `OVNKubernetes`.
+<8> The name of an existing resource group. The existing VPC and subnets should be in this resource group. The cluster is deployed to this resource group.
```
Agreed. I knew this felt familiar when we discussed it. I had actually updated published docs with the correct statement a while back. This PR has been open for so long that I had simply forgotten to update here. Good catch.

jianlinliu commented:
Go through the doc seem like the doc did not mention to update VPC's security group, actually we need it.
```
#### VPE Security Group Configuration
Traffic via the VPE Gateway needs permission within the VPC. You will need to add or update a Security Group in
the VPC to allow this traffic.

You can use the VPC's default Security Group, since that makes things easier, by adding a rule to allow traffic
from any source via TCP on port 443. You will want to make sure the VPE Gateways you setup next use this
existing or new Security Group, as that traffic will occur on 443.

Updating the default Security Group:
VPC_NAME=your-vpc-name
DEFAULT_SG=$(ibmcloud is vpc $VPC_NAME --output JSON | jq -r '.default_security_group.id')
ibmcloud is security-group-rule-add $DEFAULT_SG inbound tcp --remote 0.0.0.0/0 --port-min 443 --port-max 443
```

mjpytlak commented:
> Go through the doc seem like the doc did not mention to update VPC's security group, actually we need it.

@jianlinliu I believe it is, but want to confirm. Are you quoting IBM Cloud documentation? Asking because we already state that the VPC be configured for endpoint gateways in support of IBM Cloud Virtual Private Endpoints.  At which point, it is best practice cross reference the third-party docs. A primary reason why we typically adhere to this is we have no control over this content if/when IBM decides to update its guidance.

jianlinliu commented:
> > Go through the doc seem like the doc did not mention to update VPC's security group, actually we need it.
> 
> @jianlinliu I believe it is, but want to confirm. Are you quoting IBM Cloud documentation? Asking because we already state that the VPC be configured for endpoint gateways in support of IBM Cloud Virtual Private Endpoints. At which point, it is best practice cross reference the third-party docs. A primary reason why we typically adhere to this is we have no control over this content if/when IBM decides to update its guidance.
> 
> If you believe it important enough to mention, we can, but I think we should link to that content. If this is in fact IBM doc, please provide a link.

No, it is not from IBM Cloud documentation, but a setup internal guide from dev.

mjpytlak commented:
> Updating the default Security Group:
> VPC_NAME=your-vpc-name
> DEFAULT_SG=$(ibmcloud is vpc $VPC_NAME --output JSON | jq -r '.default_security_group.id')
> ibmcloud is security-group-rule-add $DEFAULT_SG inbound tcp --remote 0.0.0.0/0 --port-min 443 --port-max 443

@jianlinliu This seems to be shorthand that I am having trouble parsing. What are the actual commands users have to run to update the default security group?

mjpytlak commented:
> Go through the doc seem like the doc did not mention to update VPC's security group, actually we need it.

I have updated the Prerequisite section to note that the VPC must allow traffic from endpoint gateways. I have also added a new section on how to configure the VPC to allow traffic from the endpoint gateways.

jianlinliu commented:
> I have updated the Prerequisite section to note that the VPC must allow traffic from endpoint gateways. I have also added a new section on how to configure the VPC to allow traffic from the endpoint gateways.

I think it is the traffic "to" endpoint gateways, but not "from".

cjschaef commented:
>>I have updated the Prerequisite section to note that the VPC must allow traffic from endpoint gateways. I have also added a new section on how to configure the VPC to allow traffic from the endpoint gateways.

>I think it is the traffic "to" endpoint gateways, but not "from".

I think this confusion might be from some instructions I passed, regarding this Security Group (SG).

To be clear, traffic `to and from` the VPE Gateways is required.
If the user is using the default VPC SG, it is already configured with a rule to allow all outbound traffic (to the VPE's), but a rule is required to allow inbound traffic, from the VPE's, into the VPC. This is the rule that needs to be added to the default VPC SG.

Otherwise, if a new/unique SG is being configured for VPE traffic, both inbound and outbound must be configured, to my understanding.

mjpytlak commented:
> I think this confusion might be from some instructions I passed, regarding this Security Group (SG).
> 
> To be clear, traffic `to and from` the VPE Gateways is required.

Thanks for the clarification, Chris. Updated the respective sections accordingly.

jeffnowicki reviewed (changes requested):
A few more minor nits. Otherwise, it looks good.

mjpytlak commented on modules/installation-about-restricted-network.adoc:
```diff
@@ -70,6 +79,53 @@ Because of the complexity of the configuration for user-provisioned installation
+You complete the installation using a bastion host or portable device that can access both the internet and your closed network. You must use a host with internet access to:
+
+* Download the installation program, the the OpenShift CLI (`oc`), and the CCO utility (`ccoctl`).
+* Use the installation program to locate the {op-system-first} image and create the installation configuration file.
```
Not sure I agree with this update, @jeffnowicki. "Create" implies "initial". A statement about modifying the installation file before installer the cluster is technically accurate, but IMO, not necessary for this section of the docs. The purpose of this section is to let users know what they will need internet access for, not detail process.

mjpytlak commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.adoc:
```diff
@@ -0,0 +1,82 @@
+
+* You reviewed details about the xref:../../architecture/architecture-installation.adoc#architecture-installation[{product-title} installation and update] processes.
+* You xref:../../installing/installing_ibm_cloud_public/installing-ibm-cloud-account.adoc#installing-ibm-cloud-account[configured an IBM Cloud account] to host the cluster.
+* You have a container image registry that is accessible to the internet and your closed network. The container image registry should mirror the contents of the {product-registry} and contain the installation media. For more information, see xref:../../installing/disconnected_install/installing-mirroring-disconnected.adoc#installing-mirroring-disconnected[Mirroring images for a disconnected insta ...
```
Good catch. Thank you.

jianlinliu commented on modules/installation-about-restricted-network.adoc:
```diff
@@ -70,6 +79,53 @@ Because of the complexity of the configuration for user-provisioned installation
+To complete a restricted network installation, you must create a registry that
+mirrors the contents of the {product-registry} and contains the installation media.
+
+You can create this registry on a mirror host, which can access both the internet and your closed network, or by using other methods that meet your organization's security restrictions.
```
```suggestion
You can create this registry on a mirror host, which can access both the internet and your restricted network, or by using other methods that meet your organization's security restrictions.
```

cjschaef commented:
While checking the generated docs, I see this section regarding creating an install-config for restricted env's.
https://55526--ocpdocs-pr.netlify.app/openshift-enterprise/latest/installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.html#installation-initializing_installing-ibm-cloud-restricted

The steps (1 & 2), mention using the `openshift-install create install-config` command. At this time, this process requires that the IBM Cloud Account has an existing IBM Cloud CIS instance (or more), so as to prompt the user for the correct CIS domain for `baseDomain` configuration.
If the user is expecting to use IBM Cloud DNS Services, for a Private (Internal) cluster, they currently cannot rely on this command, unless they have an existing CIS instance, simply to allow the `install-config` to be generated and modify the file after, `baseDomain` and `publish: Internal`.

I'm unsure how best to provide these details, or if we need to open a bug/feature Jira to see about updating the `create install-config` command to support CIS and/or DNS Services via prompts, as those services define which publish method is being used `Public == CIS and Private == DNS Services`.

Does it make more sense to follow something similar for Private clusters?
https://docs.openshift.com/container-platform/4.14/installing/installing_ibm_cloud_public/installing-ibm-cloud-private.html#installation-initializing-manual_installing-ibm-cloud-private

Or mention the custom configuration for a Private cluster via a `NOTE` or something within this documentation, regarding using `openshift-install create install-config`?

mjpytlak commented:
> I'm unsure how best to provide these details, or if we need to open a bug/feature Jira to see about updating the `create install-config` command to support CIS and/or DNS Services via prompts,

Just to close on this @jianlinliu, @jeffnowicki, @cjschaef , and I spoke and two things are going to happen:

1. The docs will be updated to have users manually create the installation configuration file.
2. Chris will open a ticket against the installer to support DNS Services.

@cjschaef Feel free to elaborate further on the ticket, but I wanted to at least capture the decisions here.

jeffnowicki commented:
This one was 'challenging'. Nice work Mike 👍 

- lgtm
- ibm-approved

mjpytlak commented:
@jianlinliu Jeff and I have completed the final rounds of edits and he has approved the docs. Do you have additional feedback?

jianlinliu commented:
most looks good, the only comment is about the available region to run the restricted installation on IBMCloud, which is mentioned in https://github.com/openshift/openshift-docs/pull/55526#issuecomment-1880298970

For now, per my testing, only us-east region supports the VPE for all the listed private endpoints.
```
serviceEndpoints:
  - name: IAM
    url: <iam_private_endpoint_url>
  - name: VPC
    url: <vpc_private_endpoint_url>
  - name: ResourceController
    url: <resource_controller_private_endpoint_url>
  - name: ResourceManager
    url: <resource_manager_private_endpoint_url>
  - name: DNSServices
    url: <dns_services_private_endpoint_url>
  - name: COS
    url: <cos_private_endpoint_url>
  - name: GlobalSearch
    url: <global_search_private_endpoint_url>
  - name: GlobalTagging
    url: <global_tagging_private_endpoint_url>
```
So I suggest to add a NOTE for that, e.g: https://github.com/openshift/openshift-docs/pull/55526/files#diff-f4e400c4dd55ea0769811aa54453b3a4393d1db5fe828272f880b574bceb69caR385 or some other place.

mjpytlak commented:
@jianlinliu Are you referring to reinforcing the statement that VPEs are not regionally bound? If so, I have updated the docs in two places to capture this fact.

1. https://github.com/openshift/openshift-docs/pull/55526/files#diff-da2ed922fc9fc73230c8a35722f6013f310a25ef6999ca4a20bc213a75bc3b37R209
2. https://github.com/openshift/openshift-docs/pull/55526/files#diff-0995ccf569e68d8a5e211b07d6f6bd89d741f4761fb517528c10bf7b445f9593R126

jeffnowicki commented:
> most looks good, the only comment is about the available region to run the restricted installation on IBMCloud, which is mentioned in [#55526 (comment)](https://github.com/openshift/openshift-docs/pull/55526#issuecomment-1880298970)
> 
> For now, per my testing, only us-east region supports the VPE for all the listed private endpoints.
> 
> ```
> serviceEndpoints:
>   - name: IAM
>     url: <iam_private_endpoint_url>
>   - name: VPC
>     url: <vpc_private_endpoint_url>
>   - name: ResourceController
>     url: <resource_controller_private_endpoint_url>
>   - name: ResourceManager
>     url: <resource_manager_private_endpoint_url>
>   - name: DNSServices
>     url: <dns_services_private_endpoint_url>
>   - name: COS
>     url: <cos_private_endpoint_url>
>   - name: GlobalSearch
>     url: <global_search_private_endpoint_url>
>   - name: GlobalTagging
>     url: <global_tagging_private_endpoint_url>
> ```
> 
> ![](https://storage.googleapis.com/pieces-web-extensions-cdn/pieces.png)
> Copy And Save
> ![](https://storage.googleapis.com/pieces-web-extensions-cdn/link.png)
> Share
> ![](https://storage.googleapis.com/pieces-web-extensions-cdn/copilot.png)
> Ask Copilot
> ![](https://storage.googleapis.com/pieces-web-extensions-cdn/settings.png)
> So I suggest to add a NOTE for that, e.g: https://github.com/openshift/openshift-docs/pull/55526/files#diff-f4e400c4dd55ea0769811aa54453b3a4393d1db5fe828272f880b574bceb69caR385 or some other place.

@jianlinliu IBM Cloud private service endpoints (regardless of region) resolve to the same private internal network. So regardless of region you are deploying your cluster, you can pick any available private service endpoint. Selecting one in same (if available) or nearest geographic location would be preferable to minimize network latency. I'm uncertain at the moment, why your test failed.

jeffnowicki commented:
@cjschaef and I will re-evaluate services and their private service endpoints and whether there are some that have region-based restrictions.

jianlinliu commented:
> regardless of region you are deploying your cluster, you can pick any available private service endpoint. Selecting one in 
> same (if available) or nearest geographic location would be preferable to minimize network latency. I'm uncertain at the > moment, why your test failed.

@jeffnowicki @mjpytlak I think VPEs are supposed to be not regionally bound, but sounds there is some limit from IBMCloud provider. Let me take an example to make it be clear.

Assuming a user is installing a disconnected cluster in `eu-gb` region following the guide,  the user need to override `ResourceController` endpoint to use a private url -  https://private.resource-controller.cloud.ibm.com/, right? If true, how can I create the VPE (I did not see the service-controller service is listed on https://cloud.ibm.com/vpc-ext/provision/endpointGateway when  `eu-gb` region is selected)?

mjpytlak commented:
@jianlinliu - @jeffnowicki  and I had a chance to connect on our guidance around service endpoints. And after he and Chris discussed it more, Jeff requested that the latest updates be made.

jeffnowicki commented:
@jianlinliu IBM Cloud is migrating off of "Cloud (aka classic) Service Endpoints" to "Virtual Private Endpoints".  While CSE private endpoints are not necessarily regionally bound, VPE private endpoints are. With that being the case, a disconnected/restricted network installation will be constrained to those those regions that have VPE enablement for all required services. The current enabled regions are noted in the doc update @mjpytlak made. Over time, more regions will be enabled.

jianlinliu commented on installing/installing_ibm_cloud_public/installing-ibm-cloud-restricted.adoc:
```diff
@@ -0,0 +1,84 @@
++
+You cannot use the VPC that the installation program provisions by default.
+* If you plan on configuring endpoint gateways to use {ibm-cloud-name} Virtual Private Endpoints, consider the following requirements:
+** Endpoint gateway support is currently limited to the `us-east`, `us-south`, and `eu-de` regions.
```
@mjpytlak @jeffnowicki can you confirm `eu-de` region here? When I am installing a disconnected cluster in `eu-de` region, then overriding `ResourceController` serviceEndpoints to `https://private.resource-controller.cloud.ibm.com/`, but I did not find a endpoint gateway targeted to ``private.resource-controller.cloud.ibm.com` in the output of `ibmcloud is endpoint-gateway-targets --output JSON` against `eu-de` region.

jeffnowicki replied:
> can you confirm `eu-de` region here? When I am installing a disconnected cluster in `eu-de` region, then overriding `ResourceController` serviceEndpoints to `https://private.resource-controller.cloud.ibm.com/`, but I did not find a endpoint gateway targeted to ``private.resource-controller.cloud.ibm.com` in the output of`ibmcloud is endpoint-gateway-targets --output JSON` against `eu-de` region.

@jianlinliu you are correct, I confirmed that is the one service in `eu-de` that's missing a private endpoint and would be needed for a "fully disconnected" cluster installation.

@mjpytlak - please remove mention of `eu-de` from that section of the docs.

fyi @cjschaef

mjpytlak replied:
> please remove mention of `eu-de` from that section of the docs.

Removed. Thank you.

jianlinliu commented:
/lgtm

bscott-rh commented on modules/installation-ibm-cloud-configure-vpc-for-endpoint-gateways.adoc:
```diff
@@ -0,0 +1,40 @@
+
+.Prerequisites
+
+* You have installed the {ibm-cloud-name} Command Line Interface (`ibmcloud`).
```
```suggestion
* You have installed the {ibm-cloud-name} Command Line Interface utility (`ibmcloud`).
```
I'm not sure if this is new content in this PR or reused from other areas.

mjpytlak replied:
Net new. I updated it. Thanks.

bscott-rh commented on modules/installation-ibm-cloud-download-rhcos.adoc:
```diff
@@ -0,0 +1,38 @@
+[id="installation-ibm-cloud-download-rhcos_{context}"]
+= Downloading the RHCOS cluster image
+
+The installation program requires the {op-system-first} image to install the cluster. While optional, downloading the {op-system-first} before deploying, removes the need for internet access when creating the cluster.
```
```suggestion
The installation program requires the {op-system-first} image to install the cluster. While optional, downloading the {op-system-first} before deploying removes the need for internet access when creating the cluster.
```

bscott-rh commented on modules/installation-initializing-manual.adoc:
```diff
@@ -139,11 +149,113 @@ when copying installation files from an earlier {product-title} version.
++
+[NOTE]
+====
+If you choose to leave the default value of `External`, your network must be able to access the public endpoint for {ibm-cloud-name} Internet Services (CIS). CIS is not enabled for Virtual Private Endpoints.
```
```suggestion
If you use the default value of `External`, your network must be able to access the public endpoint for {ibm-cloud-name} Internet Services (CIS). CIS is not enabled for Virtual Private Endpoints.
```
minimalism nit

bscott-rh commented on modules/installation-initializing.adoc:
```diff
@@ -548,6 +558,18 @@ vpcSubnets: <vpcSubnet>
 For `platform.powervs.vpcName`, specify the name for the existing {ibm-cloud-name}. For `platform.powervs.vpcSubnets`, specify the existing subnets.
 endif::ibm-power-vs+restricted[]
+ifdef::ibm-cloud+restricted[]
+.. Define the network and subnets for the VPC to install the cluster in under the parent `platform.ibmcloud` field:
```
```suggestion
.. Define the VPC network and subnets for the cluster to use under the parent `platform.ibmcloud` field:
```
Tried rewording this because "in under" felt awkward.

mjpytlak replied:
Yeah. This is language that is used for every restricted installation assembly. I agree it reads a little awkward, but I am not sure if your suggestion changes the technical accuracy. I am inclined to leave this one alone.

bscott-rh commented on modules/installation-initializing.adoc:
```diff
@@ -573,20 +630,26 @@ publish: Internal
++
+[NOTE]
+====
+If you choose to leave the default value of `External`, your network must be able to access the public endpoint for {ibm-cloud-name} Internet Services (CIS). CIS is not enabled for Virtual Private Endpoints.
```
```suggestion
If you use the default value of `External`, your network must be able to access the public endpoint for {ibm-cloud-name} Internet Services (CIS). CIS is not enabled for Virtual Private Endpoints.
```
minimalism nit

bscott-rh reviewed (commented):
Left a few nits and suggestions, otherwise LGTM.

/remove-label peer-review-in-progress
/remove-label peer-review-needed
/label peer-review-done

mjpytlak commented on modules/installation-about-restricted-network.adoc:
```diff
@@ -70,6 +79,51 @@ Because of the complexity of the configuration for user-provisioned installation
+
+You complete the installation using a bastion host or portable device that can access both the internet and your closed network. You must use a host with internet access to:
+
+* Download the installation program, the the OpenShift CLI (`oc`), and the CCO utility (`ccoctl`).
```
Per the usual, eagle eye.

bscott-rh commented on modules/installation-about-restricted-network.adoc:
```diff
@@ -70,6 +79,51 @@ Because of the complexity of the configuration for user-provisioned installation
+If you are specifying an {ibm-name} Key Protect for {ibm-cloud-name} root key as part of the installation process, the service endpoint for Key Protect is also a required.
+====
+
+By default, the public endpoint is used to access the service. If network restrictions limit access to public service endpoints, you can override the default behavior.
```
You're right! I left this review comment before I got to the next section

mjpytlak commented on modules/installation-configuration-parameters.adoc:
```diff
@@ -2368,6 +2368,43 @@ If you are deploying the cluster into an existing VPC, the installer-provisioned
+* `ResourceController`
+* `ResourceManager`
+* `VPC`
+
```
As it stands today, there is quite a bit of variability when it comes to what a customer may choose to do when specifying an alternate endpoint. As such, we chose to be a bit more vague when it came to things like examples. As of this writing, the intent is to refine these docs over time, once virtual private endpoints and endpoint gateways become more widely available to regions.

mjpytlak commented on modules/installation-custom-ibm-cloud-vpc.adoc:
```diff
@@ -32,13 +44,35 @@ include::snippets/custom-dns-server.adoc[]
-* The name of the resource group
-* The name of VPC
-* The subnets for control plane machines and compute machines
+ifndef::ibm-cloud[]
+* The name of the existing resource group that contains the VPC and subnets
+endif::ibm-cloud[]
+ifdef::ibm-cloud[]
+* The name of the existing resource group that contains the VPC and subnets (`networkResourceGroupName`)
+endif::ibm-cloud[]
+ifndef::ibm-cloud[]
+* The name of the existing VPC
+endif::ibm-cloud[]
+ifdef::ibm-cloud[]
+* The name of the existing VPC (`vpcName`)
+endif::ibm-cloud[]
+ifndef::ibm-cloud[]
+* The subnets that were created for control plane machines and compute machines
+endif::ibm-cloud[]
+ifdef::ibm-cloud[]
+* The subnets that were created for control plane machines and compute machines (`controlPlaneSubnets` and `computeSubnets`)
+endif::ibm-cloud[]
```
I recall tinkering with this and couldn't decide if it was better to have a pile of ifndefs/idefs or go with your suggestion. I landed on this because, IMO, the published output reads easier when the parameter is on the same line of the bullet. Once this merges, I plan on coordinating with Alisha/Silke to get the Power VS equivalent parameter names. At which point I see this section as:

ifdef:ibm-cloud[]
* Item (parameter)
* Item (parameter)
* ...
endif:ibm-cloud[]

ifdef:ibm-power-vs[]
* Item (parameter)
* Item (parameter)
* ...
endif:ibm-power-vs[]

mjpytlak commented:
/cherrypick enterprise-4.15
